[nnc] Updated inliner to remove assertions and exception

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65044
* __->__ #64719

This PR also changes inlining to a 2-step process. Inlining may not always be successful. Since all mutations now happen in-place, an unsuccessful inlining transformation might leave the IR in an invalid state. To get around this problem, we clone the root stmt, try inlining on the clone, and if it succeeds, we proceed to perform inlining on the actual root stmt. This way the root stmt will always be in a valid state.

Differential Revision: [D30828583](https://our.internmc.facebook.com/intern/diff/D30828583)

navahgar committed:
[nnc] Updated inliner to remove assertions and exception

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30828583).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -667,10 +683,11 @@ class FunctionInliner : public IRMutator {
-          producer_,
-          buildErrorMessage(
-              "Producer statement for output buf should remain non-null in the fuser"));
+      if (!producer_) {
+        // Producer statement for output buf should remain non-null in the fuser
+        success_ = false;
+        return v;
+      }
```
Hm, what will happen if inliner failed somewhere on loads? Are we still going to delete the producing statement?

Since we're now performing mutations in place I wonder if this all can together lead to a partially updated, and hence incorrect, IR.

navahgar replied:
Actually, there are few complications here, which I didn't fully think through. One is that the producer statement will be mutated before the loads, so at this point there is no way to know if the inlining would succeed. Given this, the logic of returning a flag would never work.

Also, good point regarding partially successful inlining. The only way around these is to clone the IR and do what Vectorizer is doing. But then we will lose the in-place mutation property for inlining. I'm not sure which is better.

ZolotukhinM replied:
Btw, I think that my recent similar change to vectorizer might have a similar problem too, I need to double check. It would be good to split analysis and transformation parts in these passes, like good ol' guidelines from compiler development tell us :)

navahgar replied:
Yeah, it is better to check the vectorizer as well. 

We should definitely split the analysis and transformation phases. One easy way to do that in our current setup is to do what loop fusion is doing now; clone the IR, try inlining the cloned IR, if it fails return the old IR, if it succeeds, then redo the inlining on the old IR.

ZolotukhinM replied:
Right, that sounds like a sturdy solution until we implement a solid single-pass one.

navahgar replied:
Updated this PR as per this discussion. PTAL.

navahgar committed:
Update on "[nnc] Updated inliner to remove assertions and exception"


Differential Revision: [D30828583](https://our.internmc.facebook.com/intern/diff/D30828583)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30828583).

navahgar committed:
Update on "[nnc] Updated inliner to remove assertions and exception"


Differential Revision: [D30828583](https://our.internmc.facebook.com/intern/diff/D30828583)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30828583).

navahgar committed:
Update on "[nnc] Updated inliner to remove assertions and exception"


This PR also changes inlining to a 2-step process. Inlining may not always be successful. Since all mutations now happen in-place, an unsuccessful inlining transformation might leave the IR in an invalid state. To get around this problem, we clone the root stmt, try inlining on the clone, and if it succeeds, we proceed to perform inlining on the actual root stmt. This way the root stmt will always be in a valid state.

Differential Revision: [D30828583](https://our.internmc.facebook.com/intern/diff/D30828583)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30828583).

ZolotukhinM reviewed (approved):
Looks good! Thanks!

navahgar committed:
Update on "[nnc] Updated inliner to remove assertions and exception"


This PR also changes inlining to a 2-step process. Inlining may not always be successful. Since all mutations now happen in-place, an unsuccessful inlining transformation might leave the IR in an invalid state. To get around this problem, we clone the root stmt, try inlining on the clone, and if it succeeds, we proceed to perform inlining on the actual root stmt. This way the root stmt will always be in a valid state.

Differential Revision: [D30828583](https://our.internmc.facebook.com/intern/diff/D30828583)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30828583).
